src/process/script.h

```diff
@@ -0,0 +1,257 @@
+
+  if (path->path[0] == '/') {
+    ret = bpf_map_update_elem(&interpreters, new_interpreter, path, BPF_ANY);
+    if (ret < 0) {
+      error_info_t info = {0};
+      info.err = ret;
+      set_local_warning(PMW_UPDATE_MAP_ERROR, info);
+      goto EmitWarning;
+    }
+  } else {
+    relative_file_info_t rel_interpreter = {0};
+    rel_interpreter.pid = pid;
+    rel_interpreter.interpreter = *new_interpreter;
+
+    ret = bpf_map_update_elem(&rel_interpreters, &rel_interpreter, path, BPF_ANY);
+    if (ret < 0) {
+      error_info_t info = {0};
+      info.err = ret;
+      set_local_warning(PMW_UPDATE_MAP_ERROR, info);
+      goto EmitWarning;
+    }
+  }
+
+  return;
+
+ EmitWarning:;
+  process_message_t pm = {0};
+  push_warning(ctx, &pm, PM_SCRIPT);
+
+  return;
+}
+
+static __always_inline u64 push_scripts(struct pt_regs *ctx, buf_t *buffer) {
+  process_message_t *pm = (process_message_t *)buffer;
+  u64 pid_tgid = bpf_get_current_pid_tgid();
+  u32 pid = pid_tgid >> 32;
+  u64 event_id = 0;
+
+  script_t *script = bpf_map_lookup_elem(&scripts, &pid);
+  if (script == NULL) return event_id;
+
+  if (script->interpreters[0].inode == 0) goto Done;
+  // clear out the buffer
+  __builtin_memset(&pm->u.script_info.scripts, 0, sizeof(pm->u.script_info.scripts));
+
+  pm->type = PM_SCRIPT;
+  pm->u.script_info.scripts[0] = script->file.identity;
+  pm->u.script_info.buffer_length = sizeof(process_message_t);
+
+  // the original script file path came from the user; the kernel
+  // verifies that the user provided path cannot go beyond
+  // PATH_MAX. The total path can still be beyond PATH_MAX once you
+  // combine it with the CWD but userspace can take care of that
+  int sz = write_string(script->file.path, buffer, &pm->u.script_info.buffer_length, PATH_MAX);
+  if (sz < 0) goto WriteError;
+
+#pragma unroll MAX_INTERPRETERS
+  for (int i = 1; i < MAX_INTERPRETERS; i++) {
+    // there is no interpreter which means the last interpreter was
+    // the actual executable
+    if (script->interpreters[i].inode == 0) break;
+
+    // The last interpreter was a script
+    file_info_t *intp_key = &script->interpreters[i - 1];
+    pm->u.script_info.scripts[i] = *intp_key;
+
+    relative_file_info_t rel_interpreter = {0};
+    rel_interpreter.pid = pid;
+    rel_interpreter.interpreter = *intp_key;
+
+    interpreter_path_t *intp_path = bpf_map_lookup_elem(&rel_interpreters, &rel_interpreter);
+    if (intp_path == NULL) {
+      intp_path = bpf_map_lookup_elem(&interpreters, intp_key);
+    }
+
+    if (intp_path == NULL) {
+      // we couldn't find the saved path given the file identity. This
+      // most likely means we got a *lot* of interpreter scripts in
+      // quick succession and made this one be evicted from our LRU
+      // cache. This is not an error so just write a null character to
+      // tell userspace that we don't have a path for this interpreter
+      // and then move on. Do not exit early because we still got
```

Comment:
What have we still got?  The comment has left me hanging...

Reply:
Oops it's meant to be: "we still got useful information". I've updated the comment.